files/src/main/scala/io/iteratee/files/FileModule.scala

```diff
@@ -55,37 +56,41 @@ trait FileModule[F[_]] { this: Module[F] { type M[f[_]] <: MonadError[f, Throwab
   }(F)
 
   private[this] final class LineEnumerator(reader: BufferedReader) extends Enumerator[F, String] {
-    final def apply[A](step: Step[F, String, A]): F[Step[F, String, A]] =
-      if (step.isDone) F.pure(step) else F.flatMap(captureEffect(reader.readLine())) {
-        case null => F.pure(step)
-        case line => F.flatMap(step.feedEl(line))(apply)
+    final def apply[A](s: Step[F, String, A]): F[Step[F, String, A]] = F.tailRecM(s) { step =>
+      if (step.isDone) F.pure(Right(step)) else F.flatMap(captureEffect(reader.readLine())) {
+        case null => F.pure(Right(step))
+        case line => F.map(step.feedEl(line))(Left(_))
       }
+    }
   }
 
   private[this] final class ByteEnumerator(stream: InputStream, bufferSize: Int = 8192)
-    extends Enumerator[F, Array[Byte]] {
-    final def apply[A](step: Step[F, Array[Byte], A]): F[Step[F, Array[Byte], A]] =
-      if (step.isDone) F.pure(step) else F.flatten(
+      extends Enumerator[F, Array[Byte]] {
+    final def apply[A](s: Step[F, Array[Byte], A]): F[Step[F, Array[Byte], A]] = F.tailRecM(s) { step =>
+      if (step.isDone) F.pure(Right(step)) else F.flatten(
         captureEffect {
           val array = new Array[Byte](bufferSize)
           val read = stream.read(array, 0, bufferSize)
 
-          if (read == -1) F.pure(step) else F.flatMap(step.feedEl(array.slice(0, read)))(apply(_))
+          if (read == -1) F.pure(Right(step)) else F.map(step.feedEl(array.slice(0, read)))(Left(_))
         }
       )
+    }
   }
 
   private[this] final class ZipFileEnumerator(zipFile: ZipFile, iterator: Iterator[ZipEntry])
-    extends Enumerator[F, (ZipEntry, InputStream)] {
-    final def apply[A](step: Step[F, (ZipEntry, InputStream), A]): F[Step[F, (ZipEntry, InputStream), A]] =
-      if (step.isDone) F.pure(step) else F.flatten(
-        captureEffect(
-          if (iterator.hasNext) {
-            val entry = iterator.next
+      extends Enumerator[F, (ZipEntry, InputStream)] {
+    final def apply[A](s: Step[F, (ZipEntry, InputStream), A]): F[Step[F, (ZipEntry, InputStream), A]] =
+      F.tailRecM(s) { step =>
+        if (step.isDone) F.pure(Right(step)) else F.flatten(
+          captureEffect(
```

Comment:
I'll take a closer look. I originally wrote it this way since `hasNext` is observing something mutable.